system/include/emscripten/em_asm.h

```diff
@@ -66,14 +69,33 @@ void emscripten_asm_const_async_on_main_thread(
 
 // We can use the generic selection C11 feature (that clang supports pre-C11
 // as an extension) to emulate function overloading in C.
-// All pointer types should go through the default case.
+//
+// All pointer types go through the default case, which means we have different
+// defaults in `__wasm64__` and `__wasm32__` modes.
+//
+// Because the default is different we also need to include more non-default
+// cases under `__wasm32__` since, for example, we don't want `short` to end up
+// using the wider default.
+#if __wasm64__
+#define _EM_ASM_SIG_CHAR(x) _Generic((x), \
+    float: 'f', \
+    double: 'd', \
+    int: 'i', \
+    short: 'i', \
+    char: 'i', \
+    signed char: 'i', \
+    long long: 'j', \
+    unsigned: 'i', \
+    default: 'p')
```

Comment:
I find these Generic lists confusing... could we not remove all those ending in `i` for wasm32, below? And for here, why not define `long`, `unsigned long`, `unsigned long long`?

Reply:
Yes, I think we can do that.  Do you want me to try to do that before landing this?   I agree the `_Generic` is a very odd thing.. luckily this is a very limited usage .. I think we tried hard to find different ways to do this but this was the best we could do.

Reply:
See https://github.com/emscripten-core/emscripten/pull/16939

Reply:
Sounds good, thanks, but I'm still not sure why `long, unsigned long, unsigned long long` are not defined here?

Reply:
I guess maybe they should be for completeness?  I was going with what we have tests for so maybe we are lacking those?

I'll check.

Reply:
I guess in wasm64 `long` and `unsigned long` are i64, so maybe that works out, but I still worry marking them as `p` could be confusing...

`unsigned long long` seems like it has to be here because it isn't even the same type?

Reply:
I added some more tests and updated this to include all types.